### PR TITLE
CIFAR10 and CIFAR100 Datasets

### PR DESCRIPTION
**Pull request by NiklasGustafsson**
Adding CIFAR 10 and 100 to the torchvision.datasets collection.

### CODE REVIEW

File: src/Examples/SequenceToSequence.cs [hunk 1]
```diff
@@ -137,7 +137,7 @@ private static void train(int epoch, Tensor train_data, TransformerModel model,
                         total_loss += loss.to(torch.CPU).item<float>();
                     }
 
-                    if (batch % log_interval == 0 && batch > 0) {
+                    if (batch % log_interval == 0 && batch > 0 || (batch == tdlen / bptt)) {
```
Comment (tarekgh) [inline src/Examples/SequenceToSequence.cs:140]: nit: would be good if you add parentheses to these conditions to show the intention how the conditions  should be evaluated.

File: src/TorchSharp/TorchVision/dsets/CIFAR.cs [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+                    Directory.CreateDirectory(targetDir);
+                }
+
+                DownloadFile(fileName, targetDir, baseUrl);
+
+                if (!File.Exists(Path.Combine(targetDir, $"{dataset}-binary.bin"))) {
+                    DecompressFile(fileName, targetDir, targetDir);
+                }
+            }
+            private static void DecompressFile(string file, string sourceDir, string targetDir)
```
Comment (tarekgh) [inline src/TorchSharp/TorchVision/dsets/CIFAR.cs:100]: nit: insert space line here.

File: src/TorchSharp/TorchVision/dsets/CIFAR.cs [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+                Utils.Decompress.ExtractTGZ(Path.Combine(sourceDir, file), targetDir);
+            }
+
+            private void DownloadFile(string file, string target, string baseUrl)
+            {
+#if NETSTANDARD2_0_OR_GREATER
+                var filePath = NSPath.Join(target, file);
+#else
+                var filePath = Path.Join(target, file);
+#endif // NETSTANDARD2_0_OR_GREATER
```
Comment (tarekgh) [inline src/TorchSharp/TorchVision/dsets/CIFAR.cs:111]: what you think moving these into helper method to reduce the number of '#if' inserted in the sources?

File: src/TorchSharp/TorchVision/dsets/CIFAR.cs [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+                fine_labels.ForEach(d => d.Dispose());
+                coarse_labels.ForEach(d => d.Dispose());
+            }
+
+            /// <summary>
+            /// Get tensor according to index
+            /// </summary>
+            /// <param name="index">Index for tensor</param>
+            /// <returns>Tensors of index. DataLoader will catenate these tensors into batches.</returns>
+            public override Dictionary<string, Tensor> GetTensor(long index)
```
Comment (tarekgh) [inline src/TorchSharp/TorchVision/dsets/CIFAR.cs:302]: >GetTensor [](http://example.com/codeflow?start=55&length=9)

I am wondering, do we claim this is a thread safe? or we don't care?

**Review by tarekgh — DISMISSED** [review]
Added minor comments and question.

File: src/TorchSharp/TorchVision/dsets/CIFAR.cs (hunk 4 above)
Comment (NiklasGustafsson) [reply src/TorchSharp/TorchVision/dsets/CIFAR.cs:302]: No, we don't claim that. There's another PR, which is mean to make DataLoader parallel, in which we have to deal with that.

**Pull request merged**